services/net/filemonitor/FilemonitorAction.cs

```diff
@@ -0,0 +1,678 @@
+    private Dictionary<string, XmlDocument> GetXmlFileContentList(string dir, DataSourceModel dataSource)
+    {
+        var fileContentList = new Dictionary<string, XmlDocument>();
+        var fileList = Directory.GetFiles(dir);
+
+        foreach (string filePath in fileList)
+        {
+            var validDocument = GetValidXmlDocument(filePath, dataSource);
+            fileContentList.Add(filePath, validDocument);
+        }
+
+        return fileContentList;
+    }
+
+    /// <summary>
+    /// Get a list of FMS files from the import directory and return a Dictionary of strings (one per file) keyed by
+    /// the file path of each file.
+    /// </summary>
+    /// <param name="dir"></param>
+    /// <param name="dataSource"></param>
+    /// <returns></returns>
+    private Dictionary<string, string> GetFmsFileContentList(string dir, DataSourceModel dataSource)
+    {
+        var fileContentList = new Dictionary<string, string>();
+        var fileList = Directory.GetFiles(dir);
+
+        foreach (string filePath in fileList)
+        {
+            var document = ReadFileContents(filePath);
+            fileContentList.Add(filePath, document);
+        }
+
+        return fileContentList;
+    }
+    #endregion
+
+    #region Support Methods
+    /// <summary>
+    /// Get a valid XmlDocument representing the contents of the fileName parameter. Apply fixes to the XML text
+    /// in the file based on the shortcomings of the data source. Fixes are indicated by Connection string entries.
+    /// The following fixes are supported:
+    ///
+    ///     "addParent" - The BCNG files lack a single parent tag so one is inserted.
+    ///     "escapeContent" - BCNG stories contain parsing errors, so they are stored as CDATA.
+    ///
+    /// </summary>
+    /// <param name="filePath"></param>
+    /// <param name="dataSource"></param>
+    /// <returns></returns>
+    private XmlDocument GetValidXmlDocument(string filePath, DataSourceModel dataSource)
+    {
+        var xmlDoc = new XmlDocument();
+        var xmlTxt = ReadFileContents(filePath);
+
+        // BCNG files have multiple top-level objects which need to be wrapped in a single pair of tags.
+        var addParent = GetConnectionValue(dataSource, Fields.AddParent);
+        if (!String.IsNullOrEmpty(addParent))
+        {
+            if (Boolean.Parse(addParent))
+            {
+                xmlTxt = FixParentTag(xmlTxt);
+            }
+        }
+
+        // BCNG stories contain invalid XHTML content which must be escaped to parse the document.
+        var escapeContent = GetConnectionValue(dataSource, Fields.EscapeContent);
+        if (!String.IsNullOrEmpty(escapeContent))
+        {
+            if (Boolean.Parse(escapeContent))
+            {
+                xmlTxt = StoryToCdata(xmlTxt, dataSource);
+            }
+        }
+
+        try
+        {
+            xmlDoc.LoadXml(xmlTxt);
+        }
+        catch (System.Xml.XmlException e)
```

Comment:
This is an artifact of my automatic XML fixing approach. The catch block used to detect and fix the errors. I left it as-is in case this was required under a different scenario.

Reply:
`try+catch` are expensive